pallets/capacity/src/types.rs

```diff
@@ -223,3 +246,300 @@ pub fn unlock_chunks_from_vec<T: Config>(chunks: &Vec<(u32, u32)>) -> UnlockChun
+	) -> Option<usize> {
+		if self.count().is_zero() {
+			return None;
+		};
+
+		let current_staking_amount = self.get_last_staking_amount();
+		if current_staking_amount.eq(subtract_amount) && self.count().eq(&1usize) {
+			// Should not get here unless rewards have all been claimed, and provider boost history was
+			// correctly updated. This && condition is to protect stakers against loss of rewards in the
+			// case of some bug with payouts and boost history.
+			return Some(0usize);
+		}
+
+		if let Some(entry) = self.0.get_mut(reward_era) {
+			*entry = entry.saturating_sub(*subtract_amount);
+		} else {
+			self.remove_oldest_entry_if_full();
+			if self
+				.0
+				.try_insert(*reward_era, current_staking_amount.saturating_sub(*subtract_amount))
+				.is_err()
+			{
+				return None;
+			}
+		}
+		Some(self.count())
+	}
+
+	/// A wrapper for the key/value retrieval of the BoundedBTreeMap.
+	pub(crate) fn get_entry_for_era(&self, reward_era: &RewardEra) -> Option<&BalanceOf<T>> {
+		self.0.get(reward_era)
+	}
+
+	/// Returns how much was staked during the given era, even if there is no explicit entry for that era.
+	/// If there is no history entry for `reward_era`, returns the next earliest entry's staking balance.
+	///
+	/// Note there is no sense of what the current era is; subsequent calls could return a different result
+	/// if 'reward_era' is the current era and there has been a boost or unstake.
+	pub(crate) fn get_amount_staked_for_era(&self, reward_era: &RewardEra) -> BalanceOf<T> {
+		// this gives an ordered-by-key Iterator
+		let mut bmap_iter = self.0.iter();
+		let mut eligible_amount: BalanceOf<T> = Zero::zero();
+		while let Some((era, balance)) = bmap_iter.next() {
+			if era.eq(reward_era) {
+				return *balance;
+			}
+			// there was a boost or unstake in this era.
+			else if era.gt(reward_era) {
+				return eligible_amount;
+			} // eligible_amount has been staked through reward_era
+			eligible_amount = *balance;
+		}
+		eligible_amount
+	}
+
+	/// Returns the number of history items
+	pub fn count(&self) -> usize {
+		self.0.len()
+	}
+
+	fn remove_oldest_entry_if_full(&mut self) {
+		if self.is_full() {
+			// compiler errors with unwrap
+			if let Some((earliest_key, _earliest_val)) = self.0.first_key_value() {
+				self.0.remove(&earliest_key.clone());
+			}
+		}
+	}
+
+	fn get_last_staking_amount(&self) -> BalanceOf<T> {
+		// compiler errors with unwrap
+		if let Some((_last_key, last_value)) = self.0.last_key_value() {
+			return *last_value;
+		};
+		Zero::zero()
+	}
+
+	fn is_full(&self) -> bool {
+		self.count().eq(&(T::ProviderBoostHistoryLimit::get() as usize))
+	}
+}
+
+/// Struct with utilities for storing and updating unlock chunks
+#[derive(Debug, TypeInfo, PartialEqNoBound, EqNoBound, Clone, Decode, Encode, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct RetargetInfo<T: Config> {
+	/// How many times the account has retargeted this RewardEra
+	pub retarget_count: u32,
```

Comment:
can be atomic?

Reply:
Any reason why you think this needs to be atomic?

Reply:
yeah it isnt being shared and modified concurrently ,